sql/core/src/main/scala/org/apache/spark/sql/parquet/newParquet.scala

```diff
@@ -287,7 +287,11 @@ case class ParquetRelation2(
         }
       }
 
-      parquetSchema = maybeSchema.getOrElse(readSchema())
+      try {
+        parquetSchema = readSchema().getOrElse(maybeSchema.get)
+      } catch {
+        case e => throw new SparkException(s"Failed to find schema for ${paths.mkString(",")}", e)
+      }
```

Comment:
How about this 
```
parquetSchema = {
  if (maybeSchema.isDefined) {
    maybeSchema.get
  } else {
    (readSchema(), maybeMetastoreSchema) match {
      case (Some(dataSchema), _) => dataSchema
      case (None, Some(metastoreSchema)) => metastoreSchema
      case (None, None) =>
        throw new SparkException("Failed to get the schema.")
     }
  }
}
```
We first check if maybeSchema is defined. If not, we read the schema from existing data. If existing data does not exist, we are dealing with a newly created empty table and we will use maybeMetastoreSchema defined in the options.